### ISSUE TITLE
Proposal: optional full-vocabulary Q4G64 MTP proposal head (bandwidth of the draft head, coverage of the FP8 head)

### ISSUE THREAD

**Issue by splickz** [Neroued/ninfer#174]
## Summary

An optional third proposal head for `--spec mtp`: the FP8 `lm_head` requantized on the device at load into row-split Q4G64 (group-64 fp16 scales), keeping all 248,320 rows. It reads about half the bytes of the FP8 head per proposal while covering the full vocabulary, unlike the existing `--lm-head-draft` head, which is Q4G64 but only 131,072 rows plus an I32 remap. Target verification, the verification `lm_head` and the sampler are untouched; only the argmax that produces draft tokens changes.

Claim level: operator (direct GEMV benchmark) and end-to-end single-request decode. Measured on one RTX 5090 (32 GB) under WSL2, Qwen3.8-27B NVFP4 artifact, MTP5, greedy. Costs 0.63 GiB of resident memory, which makes it unsuitable for the 490K-token configuration on this card (details below).

## Why the proposal head is worth bytes

Per MTP draft, `TextContext::proposal_argmax` runs `ops::linear(hidden[5120,T], lm_head)` then `ops::argmax`, five times per MTP5 round, inside the decode graph. With `CUDA_LAUNCH_BLOCKING=1 --no-cuda-graph` the FP8 head path costs ~865 µs per proposal: GEMV ~808 µs (93%), argmax ~49 µs, memset ~4 µs, API gaps ~3 µs, two launches. That is ~4.3 ms of a 21 ms round at short context and the same ~4.3 ms of a ~38 ms round at 490K.

The FP8 GEMV is already bandwidth-bound: 1.272 GB read in 752 µs is 1.69 TB/s against the 5090's ~1.79 TB/s peak. Micro-optimizing that kernel has little room; reducing bytes is the lever. A fused GEMV+argmax was considered and not implemented, since argmax is ~6% of the proposal.

Direct operator benchmark, T=1, real dims, the production `ops::linear`/`ops::argmax` kernels, 40 timed repetitions after 5 warm-up, median:

| head | rows | linear | bytes read | effective |
|---|---:|---:|---:|---:|
| FP8_E4M3FN_ROW_BF16S (current default) | 248,320 | 752 µs | 1.272 GB | 1.69 TB/s |
| Q4G64 truncated (`--lm-head-draft`) | 131,072 | 215 µs | 0.357 GB | 1.66 TB/s |
| Q4G64 full vocabulary (proposed) | 248,320 | 402 µs | 0.675 GB | 1.68 TB/s |

Integrated (blocking attribution, same method as above) the proposal drops from ~865 µs to ~541 µs (GEMV ~474 µs, argmax ~59 µs), i.e. roughly 1.6 ms less per MTP5 round.

## Where the truncated head loses

The truncated head is the fastest option but drops any proposal whose argmax falls outside its 131,072 rows. In these runs that cost little on code and prose and a lot on structured output (same prompt, same binary, MTP5, 2 runs where shown, decode tok/s is the engine's own `req# done` figure):

| workload (short prompt) | FP8 head | truncated Q4 | full-vocab Q4 |
|---|---|---|---|
| coding, 2,088 tokens out | 228.1 tok/s, 77.5% acc | 254.3 (+11.5%), 75.5% | 248.3 (+8.9%), 77.3% |
| structured JSON, 3,500 out | 277.6, 99.0% | 278.7 (+0.4%), 86.0% | 302.5 (+9.0%), 99.0% |
| prose, 1,665 out | 105.3, 25.0% | 118.4 (+12.4%), 24.1% | 114.1 (+8.3%), 24.7% |

Accepted/attempted drafts for the structured case: 2,910/2,940 (FP8), 2,839/3,300 (truncated), 2,910/2,940 (full Q4). Generated text was byte-identical across all three heads for all three workloads, and the full-Q4 runs were byte-identical on repeat. #119 measured a similarly small on/off difference for the truncated head on code; the structured case is where the coverage matters in my data.

I am not claiming Q4 always preserves acceptance. These are one model, one card, three prompts.

## Long context (fork-only measurements)

These rows need a context beyond the native 262,144, so they were taken on the YaRN fork from #165 (release `b5f2d1a9`, upstream base `6e2786c5`), `--rope-yarn-factor 4 --kv-dtype nvfp4`. The proposal-path source is identical to master there.

| prompt | FP8 head | full-vocab Q4 | graphs-on median round (nsys) |
|---|---|---|---|
| 262K coding | 146.9 tok/s, 73.1% | 165.8 tok/s, 76.7% | 31.1 → 29.2 ms |
| 440K coding | 129.3, 75.5% | 133.8 (+3.4%), 75.7% | 36.7 → 35.3 ms |
| 490K coding | 126.3, 76.7% | 122.2 (−3.2%), 76.3% | see below |

At 262K the generated text diverged from the FP8 run about three quarters of the way through (both outputs valid code; the Q4 run stopped at 2,405 tokens, the FP8 run hit the 2,500 limit). The acceptance and tok/s difference on that row therefore compares different sequences and should not be read as an effect of the head; the median round time is the cleaner number. At 440K and 490K the outputs were byte-identical to the FP8 run.

## Memory cost and the 490K limitation

The requantized head is an additional resident allocation of 0.675 GB (free memory after startup went from 9.75 to 9.12 GiB at a 40K context). At 490K on this 32 GB card the FP8 configuration reports 31,783/32,607 MiB in use; adding the Q4 head puts the process at or past the point where WSL2/WDDM starts paging. Measured consequence: prefill fell from ~1.30k to ~0.87k tok/s, up to 17% of decode rounds stalled, and effective decode was 3% slower despite unchanged acceptance. One clean pair still showed the median round improving 37.9 → 36.2 ms, but I do not count 490K as a win: on this configuration the extra memory is the binding constraint. This is a deployment tradeoff of the added head, not an engine defect. On a card with headroom, or at a smaller context, the tradeoff does not arise; 440K on the same card was fine.

## What the change would own

- a third `ProposalHead` value and a CLI flag (`--lm-head-draft-full` in my tree, name open), and the corresponding `proposal_head` value in the request log;
- a load-time conversion FP8 row-scale → row-split Q4G64 (one kernel, one thread per row-group, scale chosen so the group's largest-magnitude element maps to −8), held for the process lifetime;
- one Q4 GEMV dispatch entry for n=248,320 (existing kernel, 40 registers, no spills);
- proposal writes into the existing full-logits buffer, so no workspace-plan change; the remap step is skipped.

Startup features, program planning and the truncated head are unchanged. An alternative would be to ship a full-vocabulary Q4 draft head inside the artifact instead of converting at load; I went with load-time conversion because it needs no artifact change, but that is the maintainer's call.

## Verification performed

Regression suite 10/10 on the patched binary (it runs without `--spec`, so it covers the unchanged paths, not the new head). Quality set of 16 short tasks: 16/16 pass on both heads, 15/16 byte-identical (one prose task differed). Needle retrieval 5/5 at a 489,803-token prompt on the fork. No NaN/Inf, CUDA or graph errors in any run. Not done: an independent numerical oracle for the requantization itself beyond the microbenchmark and the acceptance parity above; load time was not timed separately (the conversion is one pass over 19.9M groups).

Environment: RTX 5090 32 GB (SM120), WSL2 (Ubuntu 22.04) on Windows 11, CUDA 13.1, driver 610.47, `-DCMAKE_BUILD_TYPE=Release`, server command
`ninfer-serve qwen3_8_27b_nvfp4.ninfer --kv-dtype nvfp4 --max-context <ctx> --max-concurrency 1 --no-thinking --no-prefix-reuse --spec mtp --draft-tokens 5 [--lm-head-draft | --lm-head-draft-full]` (plus `--rope-yarn-factor 4 --rope-original-max-position 262144` for the long-context rows). Prompts are the coding/structured/prose fixtures from the earlier profiling harness; I can provide them and the benchmark source on request.

Would an optional head of this shape be of interest? If so I would like to agree on the flag name, whether load-time conversion or an artifact-side head is preferred, and whether the request-log value change is acceptable before touching anything further.
